File: tests/messaging/activemq-server/listener.py
import stomp
import socket
import json
import os
import time
from datetime import datetime

# Load configuration
with open('/app/config.json', 'r') as f:
    config = json.load(f)

host = config['activemq']['host']
port = config['activemq']['port']
login = config['activemq']['login']
passcode = config['activemq']['passcode']
queue = config['activemq']['queue']
ports = config['ports']
log_path = config['logPath']

# Ensure log directory exists
if not os.path.exists(log_path):
    os.makedirs(log_path)

# Log function to write incoming/outgoing messages
def log_message(direction, port, message):
    log_file = os.path.join(log_path, 'messages.log')
    timestamp = datetime.now().isoformat()
    log_entry = f"{timestamp} | {direction} | Port {port} | Message: {message}\n"
    with open(log_file, 'a') as f:
        f.write(log_entry)

# Connect to ActiveMQ
def connect_to_activemq():
    conn = stomp.Connection([(host, port)])
    conn.connect(login, passcode, wait=True)
    return conn

# Send message to ActiveMQ
def send_to_activemq(conn, port, data):
    headers = {
        'destination': queue,
        'content-type': 'text/plain',
        'port': str(port)  # Add custom property for filtering
    }
    conn.send(body=data, headers=headers)
    print(f"Sent message to ActiveMQ on port {port}")

# Subscribe to ActiveMQ
def subscribe_to_activemq(conn, port, client_socket):
    def message_listener(msg):
        body = msg.body
        log_message("Outgoing", port, body)
        client_socket.sendall(body.encode('utf-8'))

    headers = {
        'destination': queue,
        'ack': 'auto',
        'selector': f"port = '{port}'"
    }
    conn.subscribe(headers=headers, listener=message_listener)
    print(f"Subscribed to ActiveMQ on port {port}")

# Create TCP server for incoming socket connections
def create_listener(port):
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.bind(('0.0.0.0', port))
    server.listen(5)
    print(f"Listening on port {port}")

    conn = connect_to_activemq()

    while True:
        client_socket, addr = server.accept()
        print(f"Connection established on port {port} from {addr}")

        try:
            # Handle incoming data on socket
            data = client_socket.recv(1024).decode('utf-8')
            print(f"Received data on port {port}: {data}")
            log_message("Incoming", port, data)

            # Send data to ActiveMQ
            send_to_activemq(conn, port, data)

            # Subscribe to ActiveMQ and send back received messages to the client
            subscribe_to_activemq(conn, port, client_socket)

        except Exception as e:
            print(f"Error handling connection on port {port}: {e}")
        finally:
            client_socket.close()

# Main function to start the listeners for all configured ports
def main():
    for port in ports:
        create_listener(port)

if __name__ == '__main__':
    main()

File: tests/messaging/activemq-server/listener.js
const net = require('net');
const stompit = require('stompit');
const fs = require('fs');
const path = require('path');

// Load configuration
const config = JSON.parse(fs.readFileSync('/opt/jprogram/config.json', 'utf8'));
const { host, port, login, passcode, queue } = config.activemq;
const ports = config.ports;
const logPath = config.logPath;

// Ensure log directory exists
if (!fs.existsSync(logPath)) {
    fs.mkdirSync(logPath, { recursive: true });
}

const connectOptions = {
    host: host,
    port: port,
    connectHeaders: {
        'host': '/',
        'login': login,
        'passcode': passcode,
    }
};

// Log function to write incoming/outgoing messages
function logMessage(direction, port, message) {
    const logFile = path.join(logPath, 'messages.log');
    const timestamp = new Date().toISOString();
    const logEntry = `${timestamp} | ${direction} | Port ${port} | Message: ${message}\n`;
    fs.appendFileSync(logFile, logEntry);
}

// Create listener and ActiveMQ connection for each port
ports.forEach((port) => {
    // Create TCP server for incoming socket connections
    const server = net.createServer((socket) => {
        console.log(`Listening on port ${port}`);

        // Handle incoming data on socket
        socket.on('data', (data) => {
            console.log(`Received data on port ${port}: ${data}`);
            logMessage("Incoming", port, data);

            // Send the data to ActiveMQ with a 'port' property
            stompit.connect(connectOptions, (error, client) => {
                if (error) {
                    console.error(`Connection error for port ${port}:`, error.message);
                    return;
                }

                const sendHeaders = {
                    destination: queue,
                    'content-type': 'text/plain',
                    port: port  // Custom property to filter by port
                };

                const frame = client.send(sendHeaders);
                frame.write(data);
                frame.end();

                client.disconnect();
            });
        });

        // Subscribe to ActiveMQ for each port with a selector
        stompit.connect(connectOptions, (error, client) => {
            if (error) {
                console.error(`Subscription error for port ${port}:`, error.message);
                return;
            }

            const subscribeHeaders = {
                destination: queue,
                ack: 'auto',
                selector: `port = '${port}'`  // Filter messages by port property
            };

            client.subscribe(subscribeHeaders, (error, message) => {
                if (error) {
                    console.error(`Message error for port ${port}:`, error.message);
                    return;
                }

                let body = '';
                message.on('data', (chunk) => {
                    body += chunk;
                });

                message.on('end', () => {
                    logMessage("Outgoing", port, body);
                    socket.write(body); // Send back to socket
                });
            });
        });

        socket.on('end', () => {
            console.log(`Connection closed on port ${port}`);
        });
    });

    // Start server for the port
    server.listen(port, () => console.log(`Server listening on port ${port}`));
});
